README.md
# Hapi-fhir-jpaserver-oauth

[![Build Status](https://travis-ci.org/AriHealth/hapi-fhir-jpaserver-oauth.svg?branch=master)](https://travis-ci.org/AriHealth/hapi-fhir-jpaserver-oauth) [![codecov.io](https://codecov.io/gh/AriHealth/hapi-fhir-jpaserver-oauth/branch/master/graphs/badge.svg)](http://codecov.io/gh/AriHealth/hapi-fhir-jpaserver-oauth)

## Description

Open-source implementation of the FHIR specification in Java based on [HAPI](http://hapifhir.io/). HAPI FHIR CDR with support for several databases (Derby, MySQL, MariaDB and PostgreSQL) and OAuth. 


## Technology

- Java 11
- Maven for Java dependency management
- Jetty
- Spring Framework
- [HAPI](http://hapifhir.io/)

## How to deploy

Compile and package the project with:

```
mvn clean install
```

and deploy with:

```
mvn jetty:run
```

Go to your browser and type http://localhost:8080/hapi. If you enable the OAuth capability (see next section) deploy the [Keycloak OAuth 2.0](https://github.com/AriHealth/keycloak-auth) and dependencies.

## Environment variables

The following environment variables must be set prior to the execution:

* `DB_VENDOR`: Specify vendor. The list of possible vendors are [`DERBY`, `MYSQL`, `MARIADB`, `POSTGRESQL`] (`optional`, default value `DERBY`)
* `DB_HOST`: Host where the database is deployed (it can be empty for `DERBY`) (`optional`, default value `localhost`)
* `DB_PORT`: Port where the database is deployed (it can be empty for `DERBY`) (`optional`, default value `3306` for `MYSQL`, `5432` for `POSTGRESQL).
* `DB_USER`: FHIR user for the database (it can be empty for `DERBY`) (`optional`, default value `fhiruser`). 
* `DB_PASSWORD`: Password for `DB_USER` in the database (it can be empty for `DERBY`) (`optional`, default value `fhirpwd`).
* `DB_DATABASE`: DB for the database (it can be empty for `DERBY`) (`optional`, default value `fhirdb`).
* `LUCENE_FOLDER`: Place where the Lucene index are stored (`optional`, default value `/var/lib/tomcat8/webapps/hapi/indexes`).
* `OAUTH_ENABLE`: To enable/disable authentication (`optional`, default value `false`).
* `OAUTH_URL`: In case of enabling authentication, the url where the Keycloak OAuth 2.0 is available (`optional`, default value `http://auth:8081`).

## Database

This image supports `Derby`, `MySQL`, `MariaDB` and `PostgreSQL`. To configure the database use `DB_VENDOR` with the following values [`DERBY`, `MYSQL`, `MARIADB`, `POSTGRESQL`].

### Derby

No container is needed.

### MySQL
```
	docker run -d --name mysql -e MYSQL_ROOT_PASSWORD=rootpwd -e MYSQL_USER=fhiruser -e MYSQL_PASSWORD=fhirpwd MYSQL_DATABASE=fhirdb mysql:5.7
```

### MariaDB
```
	docker run -d --name mariadb -e MYSQL_ROOT_PASSWORD=rootpwd -e MYSQL_USER=fhiruser -e MYSQL_PASSWORD=fhirpwd MYSQL_DATABASE=fhirdb mariadb/server:10.3
```

### PostgreSQL
```
	docker run -d --name postgres -e POSTGRES_USER=fhiruser -e POSTGRES_PASSWORD=fhirpwd POSTGRES_DB=fhirdb postgres
```

## Docker deployment

### Full stack deployment using docker-compose

The project comes with a [docker-compose](https://docs.docker.com/compose/) which deploys testing containers for Keycloak (port: 9090), HAPI (8080), OAuth 2.0 Authenticator (8081):

0. Check the configuration values at the environment (`.env`) file. Put `OAUTH_ENABLE`=`true` to protect the API. You can modify at your needs.
1. Execute `docker-compose up -d`
2. Wait a couple of minutes until the stack is deployed. Check the logs with `docker logs --details hapi-fhir`
3. Access the Keycloak console `http://localhost:9090` (**user**: admin, **password**: Pa55w0rd)
      * Create a realm `HAPIFHIR`
      * Inside the realm create a client_id: `hapifhir-client`
      * Create a user: test. Modify the creadentials (temporary off)
4. Access the authenticator `http://localhost:8081/swagger-ui.html`
      * Open Login operation under Auth controller, `Try it out`:
      * In the JSON include the user created in Keycloak (`test`)
      * The response is a OAuth access token
      * Copy the access token
      * Last thing is to configure the [HAPI client including the authorization token in the header](http://hapifhir.io/doc_rest_client_interceptor.html). Authorization header: `Bearer <access token>` (see the java snippet below)
```
		BearerTokenAuthInterceptor authInterceptor = new BearerTokenAuthInterceptor(token); 
		// Create a client and post the transaction to the server
		IGenericClient client = ctx.newRestfulGenericClient(FHIR_URL);
		// Register the interceptor with your client (either style)
		client.registerInterceptor(authInterceptor);
```
5. Check that the resource has been created in HAPI `http://localhost:8080`

### Simple deployment

Build the image:
```
	docker build -t hapi-fhir/hapi-fhir-cdr .
```

It is supposed [auth](https://hub.docker.com/r/ccavero/keycloak-auth) is deployed. Use this command to start the container (take into account the links to auth and database containers).  For instance if MariaDB is selected the docker configuration is as follows: 
```
	docker run -d --name hapi-fhir-cdr -p 8080:8080 arihealth/hapi-fhir-cdr-oauth -e DB_VENDOR=MARIADB -e DB_HOST=mariadb -e DB_PORT=3306 -e DB_USER=fhiruser -e DB_PASSWORD=fhirpwd DB_DATABASE=fhirdb -e LUCENE_FOLDER=XXX OAUTH_ENABLE=true OAUTH_URL=http://auth:8081/ --link auth:auth --link mariadb:mariadb 
```
Note: with this command data is persisted across container restarts, but not after removal of the container. Use a new container for the database with a shared docker volume.

## MySQL configuration

We follow the recommended [MySQL configuration](https://groups.google.com/forum/#!topic/hapi-fhir/ValHrT3hAj0) including extra jpaProperties to [avoid permission problems with Lucene indexes](https://groups.google.com/forum/#!topic/hapi-fhir/wyh4TEpUuSA) of the default configuration:

1. Dependency in the pom:
```
    <dependency>
        <groupId>mysql</groupId>
        <artifactId>mysql-connector-java</artifactId>
        <version>6.0.5</version>
    </dependency>
```
2. In FhirServerConfig

* In the data source
  
```
    public DataSource dataSource() {
      BasicDataSource retVal = new BasicDataSource();
        try {
            retVal.setDriver(new com.mysql.cj.jdbc.Driver());
        } catch (SQLException e) {
        
```

* In the JPA properties

```
    private Properties jpaProperties() {

      Properties extraProperties = new Properties();

      // Use MySQL hibernate
      extraProperties.put("hibernate.dialect", "org.hibernate.dialect.MySQL5InnoDBDialect");
	  
      // To avoid problems with Lucene indexes permissions
      extraProperties.put("hibernate.search.default.indexBase", "/var/lib/tomcat8/webapps/hapi-fhir-jpaserver-example-mysql-oauth/indexes");
    }
```

## OAuth2 authorization

We use as IdM [KeyCloak](http://www.keycloak.org/). [OAuth2 authorization in HAPI](http://hapifhir.io/doc_rest_server_security.html#Authorization_Interceptor) is done [via Interceptors](http://hapifhir.io/doc_rest_server_interceptor.html). We reuse the [careconnect implementation](https://github.com/nhsconnect/careconnect-reference-implementation/blob/master/ccri-fhirgatewayhttps/src/main/java/uk/nhs/careconnect/ri/gateway/https/oauth2/OAuthTokenUtil.java) creating a new IServerInterceptor in FhirConfig that is automatically registered when launching the server:
```
    @Bean(autowire = Autowire.BY_TYPE)
    public IServerInterceptor subscriptionKeyCloakInterceptor() {
       KeyCloakInterceptor retVal = new KeyCloakInterceptor();
       return retVal;
    }
```


```
    BearerTokenAuthInterceptor authInterceptor = new BearerTokenAuthInterceptor(token);

    // Create a client and post the transaction to the server
    IGenericClient client = ctx.newRestfulGenericClient(FHIR_URL);
    // Register the interceptor with your client (either style)
    client.registerInterceptor(authInterceptor);
```


## License

Apache 2.0

By downloading this software, the downloader agrees with the specified terms and conditions of the License Agreement and the particularities of the license provided.

CHANGELOG.md
# Hapi-fhir-jpaserver-oauth

## Description

Hapi-fhir-jpaserver-oauth is HAPI version 3.8.0 with support for mySQL and OAuth. 

## Next release
### Upgrade the starter to version 3.8.0.    `bug`  
* Further information in the Pull Request [#10](https://github.com/AriHealth/hapi-fhir-jpaserver-oauth/pull/10)
### Add Travis CI to the project    `enhancement`  
* Further information in the Pull Request [#12](https://github.com/AriHealth/hapi-fhir-jpaserver-oauth/pull/12)
### Optimize the Dockerfile to separate the builder from the runner. Include java 11 support. Refine the changelog template  
* Further information in the Pull Request [#13](https://github.com/AriHealth/hapi-fhir-jpaserver-oauth/pull/13)
### Refactor packages and configuration files. Discontinue DSTU2 support and ElasticSearch instead Lucene    `enhancement`  
* Further information in the Pull Request [#14](https://github.com/AriHealth/hapi-fhir-jpaserver-oauth/pull/14)
### Provide MariaDB support for HAPI    `enhancement`  
* Further information in the Pull Request [#17](https://github.com/AriHealth/hapi-fhir-jpaserver-oauth/pull/17)
### PostgreSQL support as persistence layer    `enhancement`  
* Further information in the Pull Request [#18](https://github.com/AriHealth/hapi-fhir-jpaserver-oauth/pull/18)
### Add environment variables to configure the properties    `enhancement`  
* Further information in the Pull Request [#3](https://github.com/AriHealth/hapi-fhir-jpaserver-oauth/pull/3)
### Refactor based on sonar recommendations  
* Further information in the Pull Request [#5](https://github.com/AriHealth/hapi-fhir-jpaserver-oauth/pull/5)


src/main/java/net/atos/ari/cdr/starter/config/FhirServerConfig.java
package net.atos.ari.cdr.starter.config;

import java.sql.SQLException;
import java.util.Properties;

import javax.persistence.EntityManagerFactory;

import ca.uhn.fhir.jpa.model.entity.ModelConfig;
import ca.uhn.fhir.jpa.search.LuceneSearchMappingFactory;
import ca.uhn.fhir.jpa.util.DerbyTenSevenHapiFhirDialect;
import org.apache.commons.dbcp2.BasicDataSource;
// import org.hl7.fhir.dstu2.model.Subscription;
import org.hl7.fhir.instance.model.Subscription.SubscriptionChannelType;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowire;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.orm.jpa.JpaTransactionManager;
import org.springframework.orm.jpa.LocalContainerEntityManagerFactoryBean;
import org.springframework.transaction.annotation.EnableTransactionManagement;

import ca.uhn.fhir.jpa.config.BaseJavaConfigDstu3;
import ca.uhn.fhir.jpa.dao.DaoConfig;
import ca.uhn.fhir.jpa.util.SubscriptionsRequireManualActivationInterceptorDstu3;
import ca.uhn.fhir.rest.server.interceptor.IServerInterceptor;
import ca.uhn.fhir.rest.server.interceptor.LoggingInterceptor;
import ca.uhn.fhir.rest.server.interceptor.ResponseHighlighterInterceptor;
import net.atos.ari.cdr.starter.oauth2.KeyCloakInterceptor;

/**
 * This is the primary configuration file for the example server
 */
@Configuration
@EnableTransactionManagement()
public class FhirServerConfig extends BaseJavaConfigDstu3 {
	private static final Logger LOGGER = LoggerFactory.getLogger(FhirServerConfig.class);

	private static final String DEFAULT_MYSQL_PORT = "3306";
	private static final String DEFAULT_POSTGRESQL_PORT = "5432";
	private static final String DEFAULT_LUCENE_FOLDER = "target/lucenefiles";
	
	private static final String MARIADB_VENDOR = "MARIADB";
	private static final String MYSQL_VENDOR = "MYSQL";
	private static final String POSTGRESQL_VENDOR = "POSTGRESQL";
	private static final String DERBY_VENDOR = "DERBY";

	private static final String LOCALHOST = "localhost";

	// Const from properties
	private static final String DB_VENDOR = System.getenv("DB_VENDOR") == null? 
			DERBY_VENDOR: System.getenv("DB_VENDOR");
	private static final String DB_HOST = System.getenv("DB_HOST") == null? 
			LOCALHOST: System.getenv("DB_HOST");
	private static String DB_PORT = null;
	private static final String DB_USER = System.getenv("DB_USER") == null? 
			"": System.getenv("DB_USER");
	private static final String DB_PASSWORD = System.getenv("DB_PASSWORD") == null? 
			"": System.getenv("DB_PASSWORD");
	private static final String DB_DATABASE = System.getenv("DB_DATABASE") == null? 
			"": System.getenv("DB_DATABASE");
	private static final String LUCENE_FOLDER = System.getenv("LUCENE_FOLDER") == null? 
			DEFAULT_LUCENE_FOLDER:System.getenv("LUCENE_FOLDER");	

	/**
	 * Configure FHIR properties around the the JPA server via this bean
	 */
	@Bean()
	public DaoConfig daoConfig() {
		DaoConfig retVal = new DaoConfig();
		retVal.setAllowMultipleDelete(true);
		retVal.addSupportedSubscriptionType(SubscriptionChannelType.WEBSOCKET);
		retVal.addSupportedSubscriptionType(SubscriptionChannelType.RESTHOOK);
		retVal.addSupportedSubscriptionType(SubscriptionChannelType.EMAIL);
		retVal.setSubscriptionMatchingEnabled(true);
		return retVal;
	}

	@Bean
	public ModelConfig modelConfig() {
		return daoConfig().getModelConfig();
	} 

	/**
	 * The following bean configures the database connection. 
	 * The 'url' property value of "jdbc:derby:directory:jpaserver_derby_files;create=true"
	 * indicates that the server should save resources in a directory called "jpaserver_derby_files".
	 * 
	 * A URL to a remote database could also be placed here, along with login credentials and other properties supported by BasicDataSource.
	 * @throws SQLException 
	 */
	@Bean(destroyMethod = "close")
	public BasicDataSource dataSource() {
		BasicDataSource retVal = new BasicDataSource();
		try {

			switch (DB_VENDOR) {
				case MYSQL_VENDOR:
					retVal.setDriver(new com.mysql.jdbc.Driver());
					DB_PORT = System.getenv("DB_PORT") == null? 
							DEFAULT_MYSQL_PORT: System.getenv("DB_PORT");
					break;
				case MARIADB_VENDOR:
					retVal.setDriver(new org.mariadb.jdbc.Driver());
					DB_PORT = System.getenv("DB_PORT") == null? 
							DEFAULT_MYSQL_PORT: System.getenv("DB_PORT");
					break;
				case POSTGRESQL_VENDOR:
					retVal.setDriver(new org.postgresql.Driver());
					DB_PORT = System.getenv("DB_PORT") == null? 
							DEFAULT_POSTGRESQL_PORT: System.getenv("DB_PORT");
					break;
				case DERBY_VENDOR:
				default:
					retVal.setDriver(new org.apache.derby.jdbc.EmbeddedDriver());
					retVal.setUrl("jdbc:derby:directory:target/jpaserver_derby_files;create=true");
			}
			retVal.setUsername(DB_USER);
			retVal.setPassword(DB_PASSWORD);
			if (DB_VENDOR.equalsIgnoreCase(DERBY_VENDOR) == false)
				retVal.setUrl("jdbc:"+ DB_VENDOR.toLowerCase() +"://"+DB_HOST+":" + DB_PORT + "/" +
					DB_DATABASE + "?useSSL=false&serverTimezone=UTC");
			return retVal;
		} catch (SQLException sqlex) {
			LOGGER.error("Exception in database connection", sqlex);
			return null;
		}
	} 

	@Override
	@Bean()
	public LocalContainerEntityManagerFactoryBean entityManagerFactory() {
		LocalContainerEntityManagerFactoryBean retVal = super.entityManagerFactory();
		retVal.setPersistenceUnitName("HAPI_PU");
		retVal.setDataSource(dataSource());
		retVal.setJpaProperties(jpaProperties());
		return retVal;
	}

	private Properties jpaProperties() {
		Properties extraProperties = new Properties();
		LOGGER.info("DB_VENDOR: {}", DB_VENDOR);

		switch (DB_VENDOR) {
			case MYSQL_VENDOR:
				extraProperties.put("hibernate.dialect", "org.hibernate.dialect.MySQL5InnoDBDialect");
				break;
			case MARIADB_VENDOR:
				extraProperties.put("hibernate.dialect", "org.hibernate.dialect.MariaDB103Dialect");
				break;				
			case POSTGRESQL_VENDOR:
				extraProperties.put("hibernate.dialect", "org.hibernate.dialect.PostgreSQLDialect");
				break;				
			case DERBY_VENDOR:
			default:
				extraProperties.put("hibernate.dialect", DerbyTenSevenHapiFhirDialect.class.getName());
		}
		extraProperties.put("hibernate.format_sql", Boolean.TRUE.toString());
		extraProperties.put("hibernate.show_sql", Boolean.FALSE.toString());
		extraProperties.put("hibernate.hbm2ddl.auto", "update");
		extraProperties.put("hibernate.jdbc.batch_size", "20");
		extraProperties.put("hibernate.cache.use_query_cache", Boolean.FALSE.toString());
		extraProperties.put("hibernate.cache.use_second_level_cache", Boolean.FALSE.toString());
		extraProperties.put("hibernate.cache.use_structured_entries", Boolean.FALSE.toString());
		extraProperties.put("hibernate.cache.use_minimal_puts", Boolean.FALSE.toString());
		extraProperties.put("hibernate.search.model_mapping", LuceneSearchMappingFactory.class.getName());
		extraProperties.put("hibernate.search.default.directory_provider", "filesystem");
		extraProperties.put("hibernate.search.default.indexBase", LUCENE_FOLDER == null? DEFAULT_LUCENE_FOLDER:LUCENE_FOLDER);
		extraProperties.put("hibernate.search.lucene_version", "LUCENE_CURRENT");
		return extraProperties;
	} 

	/**
	 * Do some fancy logging to create a nice access log that has details about each incoming request.
	 */
	public LoggingInterceptor loggingInterceptor() {
		LoggingInterceptor retVal = new LoggingInterceptor();
		retVal.setLoggerName("fhirtest.access");
		retVal.setMessageFormat(
				"Path[${servletPath}] Source[${requestHeader.x-forwarded-for}] Operation[${operationType} ${operationName} ${idOrResourceName}] UA[${requestHeader.user-agent}] Params[${requestParameters}] ResponseEncoding[${responseEncodingNoDefault}]");
		retVal.setLogExceptions(true);
		retVal.setErrorMessageFormat("ERROR - ${requestVerb} ${requestUrl}");
		return retVal;
	}

	/**
	 * This interceptor adds some pretty syntax highlighting in responses when a browser is detected
	 */
	@Bean(autowire = Autowire.BY_TYPE)
	public ResponseHighlighterInterceptor responseHighlighterInterceptor() {
		return new ResponseHighlighterInterceptor();
	}

	@Bean(autowire = Autowire.BY_TYPE)
	public IServerInterceptor subscriptionSecurityInterceptor() {
		return new SubscriptionsRequireManualActivationInterceptorDstu3();
	}

	@Bean(autowire = Autowire.BY_TYPE)
	public IServerInterceptor subscriptionKeyCloakInterceptor() {
		return new KeyCloakInterceptor();
	}

	@Bean()
	public JpaTransactionManager transactionManager(EntityManagerFactory entityManagerFactory) {
		JpaTransactionManager retVal = new JpaTransactionManager();
		retVal.setEntityManagerFactory(entityManagerFactory);
		return retVal;
	}
	
}
